Added GC details to summary blobs

agarwal-navin opened the pull request:
Fixes #4534 

- Added the following data needed during GC to a new blob in summary:
  - isRootNode: This tells whether this is a root node or not.
  Root data stores have this value as true. Currently, all channel contexts have this value as true. This will change when we have root and non-root DDS.
  - gcData: The GC nodes generated by summarize.
  This information is needed when creating a new summarizer client. It needs this information from the last summary in case nothing has changed in a node (data store or DDS).

- Removed isRootDataStore from ".component" blob in data stores and moved them to the GC blob as isRootNode. It is part of GC data and can be read together with the GC nodes without needed to read attributes blob from summary.

agarwal-navin commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -620,24 +637,44 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
+             * GC details.
+             */
+            if (!(this.initSnapshotValue === null || typeof this.initSnapshotValue === "string")) {
+                const tree: ISnapshotTree = await this.initSnapshotValue;
```
As per my discussion with @arinwt, in case of differential summary, we need to decode the summary by calling decodeSummary in SummarizerNodeUtils. I will do that separately since it will need additional changes to make sure we do it in a consistent way.

arinwt replied:
Note that differential summaries are currently fully disabled by PR #4285, because of issue #4296.

anthony-murphy replied:
I'm also not convinced they should even come back fully, as they too easily hide issues. If we do keep them, i think we'll need to be much stricter with them, like forcing a document loaded via differential summary to be read-only until it sucessfully summarzies.

agarwal-navin replied:
Hmmm. I will hold on the change to handle this scenario here then. Should we create a new issue to track this? Or add a note to #4296 to handle this scenario too? @arinwt @anthony-murphy

arinwt replied:
Either way is fine but definitely add a note to that issue or add a new issue.

agarwal-navin commented on packages/runtime/datastore/src/channelContext.ts:
```diff
@@ -53,7 +54,16 @@ export function summarizeChannel(
+        isRootNode: true,
+        gcNodes: summarizeResult.gcNodes,
+    };
+    addBlobToSummary(summarizeResult, gcBlobKey, JSON.stringify(gcDetails));
```
Adding the gcNodes here means we are storing it twice since the data store will also store the nodes. This problem will be exacerbated once we have data stores within data stores. However, this is okay for now and is similar to the problem with storing outstandingOps.

vladsud replied:
Not sure I fully follow. Are you saying that we need to duplicate it in a sense of storing per-channel info, and then having aggregated data store data at level up? If so, I think that's the intent, there is no other reasonable way around it that allows pay-per-play.
But the language in you comment suggest you have a plan to address it (or maybe it's different problem), if so, having an issue to track it would be great!

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -77,21 +80,10 @@ export function createAttributesBlob(pkg: readonly string[], isRootDataStore: bo
-     * This tells whether a data store is root. Root data stores are never collected.
-     * Non-root data stores may be collected if they are not used.
-     */
-    isRootDataStore: boolean;
```
this could have N N-1 implications. probably need to keep it, and keep writing in, but stop using it, and remove in a later version

agarwal-navin replied:
Added it back.

vladsud replied:
If our back compat tests are not catching it, then likely one of the two is true:
1. It's not back compat issue (yet, not used)
2. Our tests have gap and we need to fix it (open an issie)

agarwal-navin replied:
I looked at this more carefully and it won't break back-compat. If isRooDataStore is not part of ISnapshotDetails when initializing a remote data store, it is set to true.

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -834,4 +893,8 @@ export class LocalDetachedFluidDataStoreContext
         return super.getInitialSnapshotDetails();
     }
+
+    protected async getInitialGCDetails(): Promise<IGCDetails> {
```
why override it if it just calls super?

agarwal-navin replied:
I had an assert there earlier that I removed so it became just an override :-). Removed it.

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -620,24 +639,46 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
+    protected async getInitialGCDetails(): Promise<IGCDetails> {
+        if (this.initialGcDetails === undefined) {
+            /**
+             * Default value of initial GC details in case the initial snapshot does not have GC details blob.
```
I'm not sure this comment is correct (or reflects how I understand it works).
This info is used when data store is not realized, to keep system pay-per-play.
But if old file does not have that data, we can't just use empty gcNodes, we have to realize data store and generate that info, right?

agarwal-navin replied:
Yes, thank you, I missed that! :-).
I have updated it as follows:
- If GC blob in not available (for old data stores/channel contexts), initial GC nodes are set to undefined.
- Local data store returns initial GC nodes as undefined too.
- SummarizerNodeWithGC handles the above scenarios and summarizes even if nothing has changed to get GC nodes.

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -722,8 +772,17 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
+        return {
+            isRootNode: this.isRootDataStore,
+            // Local data stores don't have initial GC nodes. These will be generated when it is summarized.
+            gcNodes: [],
```
Similar concern as above - using gcNodes = [] anywhere is incorrect (unless I missed something).
It feels like this function should not be called for local data store before first summary, and actual data should be populated and returned after any successful summary

agarwal-navin replied:
Updated as per my comment above.

agarwal-navin commented:
Ping. @vladsud @anthony-murphy. 
Note that snapshot tests are failing as expected because I added a new blob to summary. However, there is another problem where the array of GC nodes and the array out outbound routes can have their elements in different order across summaries. This is causing the tests to fail. I have added a fix for that in #4520. So that needs to go in before this PR can merge.

anthony-murphy commented:
>     gcNodes: IGraphNode[];

to solve the snapshot test issues could this be map. so gcNodes: {[id: string]: string[]}

---
Refers to: packages/runtime/runtime-definitions/src/summary.ts:41 in 5b725c4. [](commit_id = 5b725c46c72ea9e7226e59d378192cedaf04daa7, deletion_comment = False)

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -620,24 +639,44 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
+             * Default value of initial GC details in case the initial snapshot does not have GC details blob.
+             * isRootNode is set to true so that data stores in older documents are not collected by GC.
+             */
+            let gcDetails: IGCDetails = {
```
it feels like you can get rid of this variable and directly modify this.initialGcDetails

vladsud commented on packages/runtime/datastore/src/remoteChannelContext.ts:
```diff
@@ -211,4 +221,22 @@ export class RemoteChannelContext implements IChannelContext {
+                isRootNode: true,
+            };
+
+            if (await this.services.objectStorage.contains(gcBlobKey)) {
```
I think you can simply here by having
```
if (await this.services.objectStorage.contains(gcBlobKey)) {
     this.initialGcDetails = await readAndParse<IGCDetails>(this.services.objectStorage, gcBlobKey);
} else {
    this.initialGcDetails = { isRootNode: true };
}
```

vladsud commented on packages/runtime/container-runtime/src/test/dataStoreContext.spec.ts:
```diff
@@ -21,18 +31,32 @@ import { MockFluidDataStoreRuntime } from "@fluidframework/test-runtime-utils";
 } from "../dataStoreContext";
 import { ContainerRuntime } from "../containerRuntime";
 
+function getBlobFromSnapshot(blobKey: string, snapshotTree: ITree): IBlob {
```
It feels like we do the same in shipping code when working with storage, no? I.e. is there any chance of reusing existing code?

agarwal-navin replied:
I did not find anything similar. The storage APIs usually return blob from data available in storage. I didn't really want to update those.
However, I removed this function and added a simple one liner to get the blob from summary.

agarwal-navin committed:
Added blob with GC data to summary tree

agarwal-navin commented:
@anthony-murphy @vladsud Updated this PR after merging couple other GC changes.

agarwal-navin committed:
Changed tests to not use explicit function to get blob

agarwal-navin commented:
Ping @vladsud @anthony-murphy

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -662,24 +662,33 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
     }
+
+    protected async getInitialGCDetails(): Promise<IGCDetails> {
+        if (this.initialGCDetails === undefined) {
```
getInitialSnapshotDetails() is guaranteed to be called once, because it is only called by realizeCore, and it's wrapper with realize() and deferral.
isRoot(), however, can be called 20 times in a row, and each of them will result in being stuck waiting for readAndParse() below. I think what we need is for this.initialGCDetails not to be results, but promise to result, and this function not having awaits, but rather using continuations to immediately set this.initialGCDetails to pending promise.

agarwal-navin replied:
Good point!

agarwal-navin replied:
Changed this to LazyPromise which seems to fit pretty well with what we are trying to do here.
Also, since isRootDataStore is moved back to getInitialSnapshotDetails(), its no longer guaranteed that it will only be called by realize. So, I changed that to a lazy promise too.

vladsud commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -55,18 +55,18 @@ import { ContainerRuntime } from "./containerRuntime";
 
-const attributesBlobKey = ".component";
+export const attributesBlobKey = ".component";
+export const gcBlobKey = ".gc";
```
I wonder if this data should be part of .component.
The reason for this - I'm looking at perf of loading, and number of nodes (trees and blobs) in snapshot is one of the contributing factors to size of the snapshot (files like Chapter 1 has over 50% of their size in describing structure of the tree, i.e. it's proportional to number of nodes in tree). The biggest impact is DDSs (as there are more of them than data stores), but data store and .components are big factor (table 20x50 has 1002 components!)

vladsud replied:
That all said, if we assume that such blobs will be big, then it's better to have them as separate blobs, as they may be excluded from initial snapshot. With that, it's better not to name it with dot, as dot is a signal to SPO to include it in initial snapshot payload always. If it's called something else, it will be in snapshot if snapshot is small, and will not (and summarizer will load it separately on demand).

agarwal-navin replied:
These blobs can be really big. I will remove the dot from it.

vladsud commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -120,10 +122,18 @@ export class DataStores implements IDisposable {
+                 * to additional roots in the document but they won't break.
                  */
+                let isRootDataStore = true;
+                if (snapshotTree.blobs[gcBlobKey] !== undefined) {
```
If we always read this blob on boot, then I believe there is no reason for this data not to be rolled into .attributes.
We either need to do that, or separate isRoot to be in .attributes, and gc blob be only about info that only summarizer needs (i.e. can be loaded on demand)

agarwal-navin replied:
This is only for detached data stores. Remote data stores don't need to read this blob on boot. They can generate this on demand, which will only be during GC / summary.

vladsud replied:
Any way to ensure that isRoot() is not called on load path? With GC data I have less concern, as it's more obvious, but asking if it's root - I feel like we will build feature that required this on boot (for example, getRootDataStore is used today on boot, but really does not check if data store is root, as we never had that data. With this addition, it's natural to add it (once we solve some other unrelated problems in Bohemia :)).
So maybe root boolean should be part of attributes, as it is always known and part of the data store core state, but GC is something optional (to an extent).

agarwal-navin replied:
You are right. isRoot() can be called in other places as well. And it makes sense for it to be part of attributes because it seems like a part of the data store itself rather than being GC specific.

vladsud commented on packages/runtime/datastore/src/channelContext.ts:
```diff
@@ -6,12 +6,18 @@
 import { ChannelStorageService } from "./channelStorageService";
 
 export const attributesBlobKey = ".attributes";
+export const gcBlobKey = ".gc";
```
Same here - we need to either reduce number of blobs, or remove dot from it to be less important in term of inclusion into snapshot (and not be on critical boot path)

vladsud commented on packages/runtime/datastore/src/remoteChannelContext.ts:
```diff
@@ -227,8 +235,14 @@ export class RemoteChannelContext implements IChannelContext {
-        // method above.
-        return undefined;
+        if (this.initialGCDetails === undefined) {
+            if (await this.services.objectStorage.contains(gcBlobKey)) {
```
Similar feedback here = there are likely a possibility for 20 calls to come into here, there likely needs protection to only read data once (code should not assume that driver does caching if it's important for perf of this layer).

vladsud reviewed (approved):
Overall looks good, but please take action on number of blobs, naming, and impact on boot.
I.e. we either need to merge it into .attributes, or refactor not to need this blob on boot (and remove dot from name). That feels like a must before this can go in,

agarwal-navin committed:
Moved isRootDataStore back to attributes and other PR comments integration

agarwal-navin committed:
Updated test snapshot references

agarwal-navin committed:
Updated GC blob path in test

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -587,7 +597,9 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
 export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
-    private details: ISnapshotDetails | undefined;
+    private readonly intialSnapshotDetailsP =
+        new LazyPromise<ISnapshotDetails>(async () => this.loadInitialSnapshotDetails());
```
>loadInitialSnapshotDetails [](start = 59, length = 26)

i would avoid calling member function here, as they don't really makes sense as members as they should never be called elsewhere . if it's small just inline it, or move out of the class to a non-exported function

agarwal-navin replied:
These methods are not very big so I inlined them here.

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -553,6 +561,8 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
     protected abstract getInitialSnapshotDetails(): Promise<ISnapshotDetails>;
 
+    protected abstract getInitialGCDetails(): Promise<IGCDetails>;
+
```
i would remove these altogether and just take protected read-only lazy promises in the constructor

anthony-murphy replied:
or just make then properties

---
In reply to: [546138404](https://github.com/microsoft/FluidFramework/pull/4484#discussion_r546138404) [](ancestors = 546138404)

agarwal-navin replied:
For remote context, there is a lot of work happening in this method. So I am a little hesitant in making this a property for now.

agarwal-navin committed:
PR comments integration

agarwal-navin committed:
Updated remote channel context's lazy promise
